[FLINK-21949][table] Support ARRAY_AGG aggregate function

Jiabao-Sun opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

[FLINK-21949][table] Support ARRAY_AGG aggregate function

Some nosql databases like mongodb and elasticsearch support nested data types.
Aggregating multiple rows into ARRAY<ROW> is a common requirement.

## Brief change log

Introduce built in function `ARRAY_AGG([ ALL | DISTINCT ] expression)` to return an array that concatenates the input rows
and returns NULL if there are no input rows. NULL values will be ignored. Use DISTINCT for one unique instance of each value.

```sql
SELECT ARRAY_AGG(f1)
  FROM tmp
 GROUP BY f0
```

![image](https://github.com/apache/flink/assets/27403841/4ba953d0-92bf-485f-afc2-3fc292fc81ce)

Note that we have made some simplifications based on Calcite's `SqlLibraryOperators.ARRAY_AGG`.
```sql
-- calcite
ARRAY_AGG([ ALL | DISTINCT ] value [ RESPECT NULLS | IGNORE NULLS ] [ ORDER BY orderItem [, orderItem ]* ] )
-- flink
ARRAY_AGG([ ALL | DISTINCT ] expression)
```

**The differences from Calcite are as follows:**
  1. **Null values are ignored.**
  2. **The order by expression within the function is not supported because the complete row record cannot be accessed within the function implementation.**
  3. **The function returns null when there's no input rows, but calcite definition returns an empty array. The behavior was referenced from BigQuery and Postgres.**

- https://cloud.google.com/bigquery/docs/reference/standard-sql/aggregate_functions#array_agg
- https://www.postgresql.org/docs/8.4/functions-aggregate.html


## Verifying this change
ITCase and UnitCase are added.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no )
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

Jiabao-Sun commented:
Hi @wuchong, could you help review this when you have time?

snuyanzin reviewed (changes requested):
thanks for the contribution and sorry for the delay

one thing that I noticed: during aggregation there is java's hashmap is used
unfortunately java comparison rules are not fully in sync with SQL comparison rules
for that reason we can not use hash mpas here for generic cases.

We need to use custom expression evaluator, example could be seen in ArrayDistinct implementation

Jiabao-Sun commented:
Thanks @snuyanzin for the review.
Sorry, I was a little puzzled because I did not use `HashMap` in `Arrayaggfunction`.
Could you help pinpoint exactly where it is?

Jiabao-Sun commented:
Do you mean using equalityEvaluator to compare when merging and retracting lists?

Jiabao-Sun commented:
@flinkbot run azure

Jiabao-Sun commented:
Hi @snuyanzin, could you help take a look again when you have time?
Thanks.

dawidwys commented:
https://github.com/apache/flink/pull/23411#pullrequestreview-1770150649

It took me a while to understand that myself, but I think it's actually ok to depend on the `Object#equals/hashcode` in `AggregateFunctions` @snuyanzin what you said is correct for `ArrayDistinctFunction`, because:
1. it is a scalar function, so it can be chained with other functions without an exchange before it
2. It gets data from two input(s) (previous operator and e.g. a literal) haystack and needle may come from different operators which may produce data in different formats e.g. `GenericRowData` and `BinaryRowData`

In `AggregateFunctions` we will always have all records as `BinaryRowData` (and alike) and those `equals/hashcode` should work just fine. (It may be a different story when we support `STRUCTURED_TYPE#equals/hashcode`, but we will need to revisit most of the operators then, because all `MapView(s)` will work incorrectly). This is what we do in other aggregate functions already. Take a look at e.g. `JsonArrayAggFunction` or `FirstValueWithRetractAggFunction `

I haven't seen the previous version @Jiabao-Sun , but I believe we can remove the `equalityHandler` and make the `ArrayAggFunction` look similar to `JsonArrayAggFunction`.

dawidwys commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/bridging/BridgingSqlAggFunction.java:
```diff
@@ -63,6 +64,7 @@ public final class BridgingSqlAggFunction extends SqlAggFunction {
     private final List<RelDataType> paramTypes;
 
     private BridgingSqlAggFunction(
+            FlinkContext context,
```
What is the reason for those changes? Is it only to be able to create the `equalityHandler`? If so, let's keep it simple and remove those.

Jiabao-Sun replied:
Yes, I'll revert that changes.
Thanks.

snuyanzin commented:
thanks a lot for the explanation @dawidwys you're right

snuyanzin commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/ArrayAggFunctionITCase.java:
```diff
@@ -0,0 +1,95 @@
+                                Arrays.asList(
+                                        Row.of("A", new Integer[] {1, 2}),
+                                        Row.of("B", new Integer[] {2, 2, 3}),
+                                        Row.of("C", new Integer[] {3}),
```
I wonder why here is just 
```java
Row.of("C", new Integer[] {3}),
```
based on input I would expect
```java
Row.of("C", new Integer[] {3, null}),
```
or did I miss anything?

snuyanzin replied:
Also similar query for postgres
```sql
with input(a, c) as (
select 'a', 1
union all 
select 'a', 2
union all 
select 'c', 3
union all 
select 'c', null
)
select a, array_agg(distinct c) from input group by a
```
gives
```
a|array_agg|
-+---------+
a|{1,2}    |
c|{3,NULL} |
```

Jiabao-Sun replied:
```sql
-- calcite
ARRAY_AGG([ ALL | DISTINCT ] value [ RESPECT NULLS | IGNORE NULLS ] [ ORDER BY orderItem [, orderItem ]* ] )
-- flink
ARRAY_AGG([ ALL | DISTINCT ] expression)
```

This function simplifies some aspects compared to calcite. 
1. Currently, the Flink parser does not support parsing RESPECT NULLS | IGNORE NULLS, but it is still possible to make changes to support it.
2. AggregateUtil#extractDistinctInformation 848~858 ignores the ignoreNulls fields. https://github.com/apache/flink/blob/6bdb4f752adb2b43dbadd8ad4fffcb4c00568dd3/flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/AggregateUtil.scala#L848-L858
3. ListView does not support null values. https://github.com/apache/flink/blob/40cbf231e3dc15129b69136941adc15beebe8ef6/flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/dataview/ListView.java#L45-L47

For that reasons I made some simplification. If we need it, I can make some attempts.

snuyanzin replied:
It's ok to support it only partially however the main concern here: I would expect more consistent result
since there is also input `Row.ofKind(INSERT, "D", null),` and expected value `Row.of("D", null),`
So I would expect for ths either containing nulls or both not containing null however not mixed

Jiabao-Sun replied:
Thanks @snuyanzin, I'm trying to support `RESPECT NULLS | IGNORE NULLS` syntax.

Jiabao-Sun replied:
@snuyanzin 
`ListView<T>` cannot hold null values, and since it is a generic `List,` it cannot be represented by a specific object for null values. Is there any good solution for this?

https://github.com/apache/flink/blob/d92ab390ff0d91459daccfda7d6cb0acf6ca92eb/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/JsonArrayAggFunction.java#L65

dawidwys replied:
I think it's a very good point. It would be nice to be compatible with SQL and other vendors. 

One idea is we could keep the null indices and set those while retrieving the result. Happy to hear better solutions though.

snuyanzin replied:
yes, I was also thinking about storing null indexes

Jiabao-Sun replied:
Hi @dawidwys, @snuyanzin,

I refers to `LagAggFunction` use `LinkedList` to hold null values and it works. Now `RESPECT NULLS` and `IGNORE NULLS` are supported. 
https://github.com/apache/flink/blob/1ffb481111f658b699702357921a48e914d13caf/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/LagAggFunction.java#L147-L150

dawidwys replied:
The downside of the solution is that the array must fit into memory at all times. The difference with the `LagAggFunction` is that `LAG` keeps at most `n` elements where `n` is controlled by the user.

Still I am reasonably good with the `LinkedList` approach because it anyhow needs to fit into memory when we emit it at the end as a single record. Writing this down for awareness.

snuyanzin commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/ArrayAggFunctionITCase.java:
```diff
@@ -0,0 +1,95 @@
+                                Arrays.asList(
+                                        Row.of("A", new Integer[] {1, 2}),
+                                        Row.of("B", new Integer[] {2, 3}),
+                                        Row.of("C", new Integer[] {3}),
```
same here 
based on input I would expect 
```java
Row.of("C", new Integer[] {3, null}),
```

Jiabao-Sun commented:
Thanks @snuyanzin @dawidwys for the review.
Could you help review it again?

Jiabao-Sun commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/ArrayAggFunctionITCase.java:
```diff
@@ -0,0 +1,95 @@
+                TestSpec.forFunction(BuiltInFunctionDefinitions.ARRAY_AGG)
+                        .withDescription("ARRAY changelog stream aggregation")
+                        .withSource(
+                                ROW(STRING(), INT()),
```
Full types test is in `ArrayAggFunctionTest`.

snuyanzin commented:
Thanks for addressing comments 
in general it looks ok from my side
i guess there is one little thing: since it is based on Calcite parser it allows to have `ORDER BY` inside...
At the same time it is currently not supported on Flink level, not sure whether we can redefine this behavior however at least it would make sense to mention it in doc that it is not supported

Jiabao-Sun commented:
> Thanks for addressing comments in general it looks ok from my side i guess there is one little thing: since it is based on Calcite parser it allows to have `ORDER BY` inside... At the same time it is currently not supported on Flink level, not sure whether we can redefine this behavior however at least it would make sense to mention it in doc that it is not supported

Yes, ORDER BY allows sorting of any field in the input rows, but currently it is difficult to obtain the complete input rows for sorting in the function implementation. Therefore, the `ORDER BY` clause is not supported yet.
I have added an explanation in the documentation.

@snuyanzin, please help take a look again when you have time.

Jiabao-Sun commented:
@flinkbot run azure

snuyanzin commented on docs/data/sql_functions.yml:
```diff
@@ -1059,6 +1059,13 @@ aggregate:
+    description: |
+      By default or with keyword `ALL` and, return an array that concatenates the input rows
+      and returns `NULL` if there are no input rows. Use `DISTINCT` for one unique instance of each value.
+      By default null values are respected, use `IGNORE NULLS` to skip null values.
```
```suggestion
      By default `NULL` values are respected, use `IGNORE NULLS` to skip null values.
```

snuyanzin reviewed (approved):
Looks ok from my side

@dawidwys do you have anything to add (asking since you're also participating here)

Jiabao-Sun commented:
@flinkbot run azure

Jiabao-Sun commented:
Hi @dawidwys, please help review it again when you have time.
Thanks a lot.

dawidwys commented:
Sorry, it takes such a long time from my side. I had a vacation in the meantime. I'll try to check it Monday. Nevertheless if you're comfortable with the PR @snuyanzin feel free to merge it without waiting for my review.

snuyanzin commented:
I think we could wait until Monday or even more since right now a feature freeze and need to wait for cutting release branch

dawidwys commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/ArrayAggFunction.java:
```diff
@@ -0,0 +1,194 @@
+                buffer.add(element);
+            }
+            // merge retract list of acc and other
+            List<T> retractBuffer = new ArrayList<>();
+            for (T element : acc.retractList) {
+                retractBuffer.add(element);
+            }
+            for (T element : otherAcc.retractList) {
+                retractBuffer.add(element);
+            }
```
do we need the `retractBuffer`? Can't we just iterate over both the `retractList` and create only the final `newRetractBuffer`? It seems we create quite some unnecessary objects and potentially list resizing here.

dawidwys commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/ArrayAggFunction.java:
```diff
@@ -0,0 +1,194 @@
+
+            // update to acc
+            acc.list.clear();
+            acc.list.addAll(buffer);
```
can't we populate `acc.list` in a single go? The current approach does make sense with a `ListView` but it does not with a `LinkedList` kept in memory.

I believe we don't need the intermediate `buffer` and `newRetractBuffer`

dawidwys commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/ArrayAggFunction.java:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    public void retract(ArrayAggAccumulator<T> acc, T value) throws Exception {
+        if (value != null) {
```
what about retracting nulls?

dawidwys replied:
would be super nice to have a test case for that

dawidwys reviewed (changes requested):
I put some comments to the implementation

dawidwys commented:
Thanks for the update @Jiabao-Sun The implementation looks good now. I want to go through the tests again, but I need a bit more time. I hope this is fine, cause anyway we need to wait for a branch cut.

dawidwys reviewed (approved):
LGTM, thank you for the contribution @Jiabao-Sun !
